### PR TITLE
drivers: sensors: Add support for RaspberryPi Pico CPU temperature

### PR DESCRIPTION
**Pull request by soburi**
Support for the measuring the CPU die temperature
for the RaspberryPi Pico.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

### CODE REVIEW

File: drivers/sensor/rpi_pico_temp/Kconfig [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# Raspberry Pi Pico temperature sensor configuration options
+
+# Copyright (c) 2023 TOKITA Hiroshi
+# SPDX-License-Identifier: Apache-2.0
+
+config RPI_PICO_TEMP
+	bool "Raspberry Pi Pico CPU Temperature Sensor"
+	default y
+	depends on DT_HAS_RASPBERRYPI_PICO_TEMP_ENABLED
+	select ADC
```
Comment (yonsch) [inline drivers/sensor/rpi_pico_temp/Kconfig:10]: Should this select ADC, or rather depend on ADC? stm32 temp uses depends on...

File: dts/bindings/sensor/raspberrrypi,pico-temp.yaml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2023 TOKITA Hiroshi
+# SPDX-License-Identifier: Apache-2.0
+
+description: RaspberryPi Pico family TEMP node
```
Comment (yonsch) [inline dts/bindings/sensor/raspberrrypi,pico-temp.yaml:4]: Maybe expand TEMP to "temperature sensor"

File: dts/arm/rpi_pico/rp2040.dtsi [hunk 3]
```diff
@@ -185,6 +185,13 @@
 	pinctrl: pin-controller {
 		compatible = "raspberrypi,pico-pinctrl";
 	};
+
+	coretemp: coretemp {
+		compatible = "raspberrypi,pico-temp";
+		io-channels = <&adc 4>;
+		status = "okay";
+	};
+
```
Comment (yonsch) [inline dts/arm/rpi_pico/rp2040.dtsi:194]: Extra blank line

File: dts/arm/rpi_pico/rp2040.dtsi [hunk 4]
```diff
@@ -185,6 +185,13 @@
 	pinctrl: pin-controller {
 		compatible = "raspberrypi,pico-pinctrl";
 	};
+
+	coretemp: coretemp {
```
Comment (yonsch) [inline dts/arm/rpi_pico/rp2040.dtsi:189]: Maybe call this die-temp like the channel?

File: drivers/sensor/rpi_pico_temp/rpi_pico_temp.c [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	struct rpi_pico_temp_data *data = dev->data;
+	int rc;
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_DIE_TEMP) {
+		return -ENOTSUP;
+	}
+
+	rc = adc_channel_setup(cfg->adc, &cfg->ch_cfg);
+	if (rc) {
+		LOG_DBG("Setup AIN%u got %d", cfg->ch_cfg.channel_id, rc);
```
Comment (yonsch) [inline drivers/sensor/rpi_pico_temp/rpi_pico_temp.c:41]: AIN should be replaced by a more pico suitable term

File: drivers/sensor/rpi_pico_temp/rpi_pico_temp.c [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+	if (rc) {
+		LOG_DBG("adc_raw_to_millivolts() failed %d", rc);
+		return rc;
+	}
+
+	/*
+	 * Calculate CPU temperature from voltage by the equation:
+	 * T = 27 - (ADC_Voltage - 0.706)/0.001721
+	 */
+	work = ((27 * 1721) - (mv * 1000 - 706000));
```
Comment (yonsch) [inline drivers/sensor/rpi_pico_temp/rpi_pico_temp.c:75]: As the datasheet states:

> As the Vbe and the Vbe slope can vary over the temperature range, and from device to device, some user calibration may be required if accurate measurements are required.

So, I think these constants should come from DTS, with the values you've used as the defaults.

**Review by yonsch — CHANGES_REQUESTED** [review]
Thanks, left a few comments

File: drivers/sensor/rpi_pico_temp/Kconfig (hunk 1 above)
Comment (soburi) [reply drivers/sensor/rpi_pico_temp/Kconfig]: Follows the manners in stm32.
Change it to `depends ADC`

File: dts/bindings/sensor/raspberrrypi,pico-temp.yaml (hunk 2 above)
Comment (soburi) [reply dts/bindings/sensor/raspberrrypi,pico-temp.yaml]: Corrected as your pointing out.

File: dts/arm/rpi_pico/rp2040.dtsi (hunk 3 above)
Comment (soburi) [reply dts/arm/rpi_pico/rp2040.dtsi]: Removed the line.

File: drivers/sensor/rpi_pico_temp/rpi_pico_temp.c (hunk 5 above)
Comment (soburi) [reply drivers/sensor/rpi_pico_temp/rpi_pico_temp.c]: Change it to `ADC channel`

File: drivers/sensor/rpi_pico_temp/rpi_pico_temp.c (hunk 6 above)
Comment (soburi) [reply drivers/sensor/rpi_pico_temp/rpi_pico_temp.c]: Added defines `vbe` and `vbe-slope` to the node.

File: samples/sensor/cpu_temp_polling/boards/rpi_pico.overlay [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+/ {
+	aliases {
+		cputemp0 = &coretemp;
+	};
+};
+
+&coretemp {
+	status = "okay";
+};
```
[the remark this answers is not in the document]
Comment (soburi) [reply samples/sensor/cpu_temp_polling/boards/rpi_pico.overlay]: `aliase` clause move to rp2040.dts.
Still default disabled the `die-temp` node.
So keeping remain the overlay for enabling it.

File: drivers/sensor/rpi_pico_temp/Kconfig [hunk 8]
```diff
@@ -0,0 +1,13 @@
+# Raspberry Pi Pico temperature sensor configuration options
+
+# Copyright (c) 2023 TOKITA Hiroshi
+# SPDX-License-Identifier: Apache-2.0
+
+config RPI_PICO_TEMP
+	bool "Raspberry Pi Pico CPU Temperature Sensor"
+	default y
+	depends on DT_HAS_RASPBERRYPI_PICO_TEMP_ENABLED
+	depends on (ADC && SOC_SERIES_RP2XXX)
```
Comment (MaureenHelm) [inline drivers/sensor/rpi_pico_temp/Kconfig:10]: You already have a dependency on the dts compatible, so I think you can remove the redundant dependency on the SOC_SERIES

File: dts/bindings/sensor/raspberrrypi,pico-temp.yaml [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+include: sensor-device.yaml
+
+properties:
+  io-channels:
+    required: true
+    description: ADC channel for CPU temperature sensor
+
+  vbe:
+    type: int
+    default: 706000
```
Comment (MaureenHelm) [inline dts/bindings/sensor/raspberrrypi,pico-temp.yaml:17]: Please justify the default value in the description

File: dts/bindings/sensor/raspberrrypi,pico-temp.yaml [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+    description: |
+      Vbe voltage in microvolts at 27 degree C.
+      The temperature sensor measures the Vbe voltage of a biased bipolar diode.
+
+  vbe-slope:
+    type: int
+    default: -1721
```
Comment (MaureenHelm) [inline dts/bindings/sensor/raspberrrypi,pico-temp.yaml:24]: Please justify the default value in the description

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

**Review by yonsch — DISMISSED** [review]

File: drivers/sensor/rpi_pico_temp/Kconfig (hunk 8 above)
Comment (soburi) [reply drivers/sensor/rpi_pico_temp/Kconfig]: It make a sense.
Removed SOC_SERIES_RP2XXX.

File: dts/bindings/sensor/raspberrrypi,pico-temp.yaml (hunk 9 above)
Comment (soburi) [reply dts/bindings/sensor/raspberrrypi,pico-temp.yaml:17]: Add short description for clarify it is a spec.

File: dts/bindings/sensor/raspberrrypi,pico-temp.yaml (hunk 10 above)
Comment (soburi) [reply dts/bindings/sensor/raspberrrypi,pico-temp.yaml:25]: Add short description for clarify it is a spec.

**Review by yonsch — DISMISSED** [review]

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Can you please rename the dts alias and sample to be consistent with the sensor channel name? i.e., die temp instead of cpu temp

**Comment by soburi** [comment]
@MaureenHelm 
> Can you please rename the dts alias and sample to be consistent with the sensor channel name? i.e., die temp instead of cpu temp

Renamed to die-temp[n].

**Review by yonsch — DISMISSED** [review]

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
> > Can you please rename the dts alias and sample to be consistent with the sensor channel name? i.e., die temp instead of cpu temp
> 
> Renamed to die-temp[n].

The sample filenames need to be renamed too

**Comment by soburi** [comment]
@MaureenHelm 
> The sample filenames need to be renamed too
Fix done.

**Review by yonsch — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**